File: exercism/fsharp/bob/Bob.fs
module Bob

let (|Shouting|_|) (interaction: string) =
  if (interaction.ToUpper() = interaction && interaction.ToLower() <> interaction) then Some() else None

let (|Question|_|) (interaction: string) =
  if (interaction.EndsWith("?")) then Some() else None

let (|Silence|_|) (interaction: string) =
  if (interaction = "") then Some() else None

let (|Other|) _ = Other

let response (interaction: string): string =
  match interaction.Trim() with
  | Question & Shouting -> "Calm down, I know what I'm doing!"
  | Question -> "Sure."
  | Shouting -> "Whoa, chill out!"
  | Silence -> "Fine. Be that way!"
  | Other -> "Whatever."

File: exercism/clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as s]))

(defn silence? [s]
  (= "" (s/trim s)))

(defn yelling? [s]
  (and
    (= s (s/upper-case s))
    (not= s (s/lower-case s))))

(defn question? [s]
  (s/ends-with? s "?"))

(def forceful-question? (every-pred yelling? question?))

(defn response-for [s]
  (cond
    (silence? s) "Fine. Be that way!"
    (forceful-question? s) "Calm down, I know what I'm doing!"
    (yelling? s) "Whoa, chill out!"
    (question? s) "Sure."
    :else "Whatever."))
